doc/musig-reference.py

```diff
@@ -0,0 +1,402 @@
+    if P1 is None:
+        return P2
+    if P2 is None:
+        return P1
+    if (x(P1) == x(P2)) and (y(P1) != y(P2)):
+        return None
+    if P1 == P2:
+        lam = (3 * x(P1) * x(P1) * pow(2 * y(P1), p - 2, p)) % p
+    else:
+        lam = ((y(P2) - y(P1)) * pow(x(P2) - x(P1), p - 2, p)) % p
+    x3 = (lam * lam - x(P1) - x(P2)) % p
+    return (x3, (lam * (x(P1) - x3) - y(P1)) % p)
+
+def point_mul(P: Optional[Point], n: int) -> Optional[Point]:
+    R = None
+    for i in range(256):
+        if (n >> i) & 1:
+            R = point_add(R, P)
+        P = point_add(P, P)
+    return R
+
+def bytes_from_int(x: int) -> bytes:
+    return x.to_bytes(32, byteorder="big")
+
+def bytes_from_point(P: Point) -> bytes:
+    return bytes_from_int(x(P))
+
+def lift_x(b: bytes) -> Optional[Point]:
+    x = int_from_bytes(b)
+    if x >= p:
+        return None
+    y_sq = (pow(x, 3, p) + 7) % p
+    y = pow(y_sq, (p + 1) // 4, p)
+    if pow(y, 2, p) != y_sq:
+        return None
+    return (x, y if y & 1 == 0 else p-y)
+
+def int_from_bytes(b: bytes) -> int:
+    return int.from_bytes(b, byteorder="big")
+
+def has_even_y(P: Point) -> bool:
+    assert not is_infinite(P)
+    return y(P) % 2 == 0
+
+#
+# End of helper functions copied from BIP-340 reference implementation.
+#
+
+infinity = None
+
+def cbytes(P: Point) -> bytes:
+    a = b'\x02' if has_even_y(P) else b'\x03'
+    return a + bytes_from_point(P)
+
+def point_negate(P: Point) -> Point:
+    if is_infinite(P):
+        return P
+    return (x(P), p - y(P))
+
+def pointc(x: bytes) -> Point:
+    P = lift_x(x[1:33])
+    if x[0] == 2:
+        return P
+    elif x[0] == 3:
+        return point_negate(P)
+    assert False
+
+def key_agg(pubkeys: List[bytes], tweaks: List[bytes], is_xonly: List[bool]) -> bytes:
+    Q, _, _ = key_agg_internal(pubkeys, tweaks, is_xonly)
+    return bytes_from_point(Q)
+
+def key_agg_internal(pubkeys: List[bytes], tweaks: List[bytes], is_xonly: List[bool]) -> Point:
+    pk2 = get_second_key(pubkeys)
+    u = len(pubkeys)
+    Q = infinity
+    for i in range(u):
+        P_i = lift_x(pubkeys[i])
+        a_i = key_agg_coeff_(pubkeys, pubkeys[i], pk2)
+        Q = point_add(Q, point_mul(P_i, a_i))
+    assert not is_infinite(Q)
+    gacc = 1
+    tacc = 0
+    v = len(tweaks)
+    for i in range(v):
+        Q, gacc, tacc = apply_tweak(Q, gacc, tacc, tweaks[i], is_xonly[i])
+    return Q, gacc, tacc
+
+def hash_keys(pubkeys: List[bytes]) -> bytes:
+    return tagged_hash('KeyAgg list', b''.join(pubkeys))
+
+def get_second_key(pubkeys: List[bytes]) -> bytes:
+    u = len(pubkeys)
+    for j in range(1, u):
+        if pubkeys[j] != pubkeys[0]:
+            return pubkeys[j]
+    return bytes_from_int(0)
+
+def key_agg_coeff(pubkeys: List[bytes], pk_: bytes) -> int:
+    pk2 = get_second_key(pubkeys)
+    return key_agg_coeff_(pubkeys, pk_, pk2)
+
+def key_agg_coeff_(pubkeys: List[bytes], pk_: bytes, pk2: bytes) -> int:
```

Comment:
nit: Should we rename this to `key_agg_coeff_internal` then?

Reply:
Yes, done!